### PR TITLE
HBASE-23078 BaseLoadBalancer should consider region replicas when ran…

### PR DESCRIPTION
**Pull request by infraio**
…domAssignment and roundRobinAssignment

### CODE REVIEW

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java [hunk 1]
```diff
@@ -1319,8 +1320,22 @@ protected boolean areSomeRegionReplicasColocated(Cluster c) {
...
-      boolean hasRegionReplica) {
+  protected Cluster createCluster(List<ServerName> servers, Collection<RegionInfo> regions)
+      throws HBaseIOException {
+    boolean hasRegionReplica = false;
+    try {
+      for (RegionInfo regionInfo : regions) {
+        TableName tableName = regionInfo.getTable();
+        if (services != null &&
+            services.getTableDescriptors().get(tableName).getRegionReplication() > 1) {
+          hasRegionReplica = true;
```
Comment (saintstack) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java]: Good

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m 12s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 56s |  master passed  |
| :green_heart: |  compile  |   0m 55s |  master passed  |
| :green_heart: |  checkstyle  |   1m 33s |  master passed  |
| :green_heart: |  shadedjars  |   5m  9s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 38s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 30s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 27s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 31s |  the patch passed  |
| :green_heart: |  compile  |   0m 59s |  the patch passed  |
| :green_heart: |  javac  |   0m 59s |  the patch passed  |
| :broken_heart: |  checkstyle  |   1m 28s |  hbase-server: The patch generated 2 new + 61 unchanged - 0 fixed = 63 total (was 61)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m  3s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m 25s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 37s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 40s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 234m 32s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 30s |  The patch does not generate ASF License warnings.  |
|  |   | 297m 34s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.TestRegionPlacement2 |
|   | hadoop.hbase.master.balancer.TestBaseLoadBalancer |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/663 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 0460cea06f1d 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-663/out/precommit/personality/provided.sh |
| git revision | master / f0dddd1cc2 |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/1/artifact/out/diff-checkstyle-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/1/testReport/ |
| Max. process+thread count | 4963 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java [hunk 2]
```diff
@@ -1319,8 +1318,22 @@ protected boolean areSomeRegionReplicasColocated(Cluster c) {
...
-  protected Cluster createCluster(List<ServerName> servers, Collection<RegionInfo> regions,
-      boolean hasRegionReplica) {
+  protected Cluster createCluster(List<ServerName> servers, Collection<RegionInfo> regions)
+      throws HBaseIOException {
+    boolean hasRegionReplica = false;
+    try {
+      for (RegionInfo regionInfo : regions) {
+        TableName tableName = regionInfo.getTable();
+        if (services != null &&
+            services.getTableDescriptors().get(tableName).getRegionReplication() > 1) {
```
Comment (ramkrish86) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java]: I think the descriptors are anyway cached right? If so good.

**Comment by ramkrish86** [comment]
Looks good to me.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   2m 28s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   8m  3s |  master passed  |
| :green_heart: |  compile  |   1m 14s |  master passed  |
| :green_heart: |  checkstyle  |   1m 40s |  master passed  |
| :green_heart: |  shadedjars  |   6m 15s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 52s |  master passed  |
| :blue_heart: |  spotbugs  |   5m 22s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   5m 20s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   7m 27s |  the patch passed  |
| :green_heart: |  compile  |   1m 18s |  the patch passed  |
| :green_heart: |  javac  |   1m 18s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 38s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   6m 16s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  23m  2s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 48s |  the patch passed  |
| :green_heart: |  findbugs  |   6m 27s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 240m 23s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 26s |  The patch does not generate ASF License warnings.  |
|  |   | 322m 18s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.TestRegionPlacement2 |
|   | hadoop.hbase.master.TestMasterShutdown |
|   | hadoop.hbase.master.balancer.TestBaseLoadBalancer |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/663 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux b9e768eaa24c 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-663/out/precommit/personality/provided.sh |
| git revision | master / 3250a8021a |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/2/testReport/ |
| Max. process+thread count | 5097 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 35s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 26s |  master passed  |
| :green_heart: |  compile  |   0m 56s |  master passed  |
| :green_heart: |  checkstyle  |   1m 21s |  master passed  |
| :green_heart: |  shadedjars  |   4m 31s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 38s |  master passed  |
| :blue_heart: |  spotbugs  |   4m  1s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   3m 59s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   4m 55s |  the patch passed  |
| :green_heart: |  compile  |   0m 56s |  the patch passed  |
| :green_heart: |  javac  |   0m 56s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 19s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 32s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  15m 30s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 10s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 157m 12s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 35s |  The patch does not generate ASF License warnings.  |
|  |   | 213m 34s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/663 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 44d0e592d884 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-663/out/precommit/personality/provided.sh |
| git revision | master / 3250a8021a |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/3/testReport/ |
| Max. process+thread count | 4855 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java (hunk 2 above)
Comment (infraio) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/BaseLoadBalancer.java]: Updated in new patch. Will only getTableDescriptors once now.

**Comment by infraio** [comment]
Add the region replica UT back.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   2m 10s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   7m 43s |  master passed  |
| :green_heart: |  compile  |   1m  2s |  master passed  |
| :green_heart: |  checkstyle  |   1m 35s |  master passed  |
| :green_heart: |  shadedjars  |   5m 21s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 41s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 40s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 37s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 39s |  the patch passed  |
| :green_heart: |  compile  |   1m  4s |  the patch passed  |
| :green_heart: |  javac  |   1m  4s |  the patch passed  |
| :broken_heart: |  checkstyle  |   1m 34s |  hbase-server: The patch generated 1 new + 71 unchanged - 0 fixed = 72 total (was 71)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m 17s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  18m 20s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 34s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 32s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 244m 16s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 28s |  The patch does not generate ASF License warnings.  |
|  |   | 312m  1s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.TestMasterShutdown |
|   | hadoop.hbase.client.TestAsyncRegionAdminApi2 |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/663 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux ce172c12f057 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-663/out/precommit/personality/provided.sh |
| git revision | master / ef79b40e6e |
| Default Java | 1.8.0_181 |
| checkstyle | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/4/artifact/out/diff-checkstyle-hbase-server.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/4/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/4/testReport/ |
| Max. process+thread count | 5062 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/4/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m 15s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 59s |  master passed  |
| :green_heart: |  compile  |   0m 57s |  master passed  |
| :green_heart: |  checkstyle  |   1m 32s |  master passed  |
| :green_heart: |  shadedjars  |   5m  5s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 40s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 25s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 23s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 30s |  the patch passed  |
| :green_heart: |  compile  |   0m 56s |  the patch passed  |
| :green_heart: |  javac  |   0m 56s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 28s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m  3s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m  9s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 35s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 32s |  the patch passed  |
||| _ Other Tests _ |
| :broken_heart: |  unit  | 254m  8s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 26s |  The patch does not generate ASF License warnings.  |
|  |   | 316m 28s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.assignment.TestRegionBypass |
|   | hadoop.hbase.util.TestFromClientSide3WoUnsafe |
|   | hadoop.hbase.master.procedure.TestRestoreSnapshotProcedure |
|   | hadoop.hbase.master.TestSplitWALManager |
|   | hadoop.hbase.master.assignment.TestMergeTableRegionsProcedure |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/5/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/663 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 672d38183af9 4.15.0-54-generic #58-Ubuntu SMP Mon Jun 24 10:55:24 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-663/out/precommit/personality/provided.sh |
| git revision | master / 836368189a |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/5/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/5/testReport/ |
| Max. process+thread count | 5006 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-663/5/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by infraio** [comment]
The failed ut not related and passed.

**Pull request merged**